### PR TITLE
🔨 switch almost all remaining typeorm util function holdouts to knex

### PR DESCRIPTION
**Pull request by danyx23**
This PR cleans up almost all remaining uses of TypeORM utility funcitons. The only remaining holdouts now are the ones in the chart revision tool related functions which will be done in a later PR.

### CODE REVIEW

**Comment by danyx23** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3359?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3359** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3359?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#3358** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3358?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3299** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3299?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3298** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3298?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3293** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3293?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3292** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3292?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3271** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3259** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3259?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#3258** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3258?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @danyx23 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: adminSiteServer/plainRouterHelpers.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import { NextFunction, Request, Response, Router } from "express"
+import * as db from "../db/db.js"
+export function getPlainRouteWithROTransaction<T>(
```
Comment (ikesau) [inline adminSiteServer/plainRouterHelpers.tsx:3]: I wasn't even aware of the plain/functional router distinction!

Not for the PR of course, but do you think it's worth keeping both around?

File: adminSiteServer/testPageRouter.tsx [hunk 2]
```diff
@@ -437,19 +438,23 @@ function EmbedTestPage(props: EmbedTestPageProps) {
     )
 }
 
-testPageRouter.get("/embeds", async (req, res) => {
-    const props = await db.knexReadonlyTransaction((trx) =>
-        propsFromQueryParams(trx, {
+getPlainRouteWithROTransaction(
```
Comment (ikesau) [inline adminSiteServer/testPageRouter.tsx:441]: At http://localhost:3030/admin/test , only the explorer test embeds appear to work

Everything else gets:
```
TypeError: The operator "undefined" is not permitted
```

File: baker/SiteBaker.tsx [hunk 3]
```diff
@@ -431,8 +431,6 @@ export class SiteBaker {
     private async removeDeletedPosts(knex: db.KnexReadonlyTransaction) {
         if (!this.bakeSteps.has("removeDeletedPosts")) return
 
-        await db.getConnection()
```
Comment (ikesau) [inline baker/SiteBaker.tsx:434]: I don't think this change has come from here, but I also checked in the head of this PR chain and it's happening there too so I'll just write it here that:

```
node itsJustJavascript/baker/buildLocalBake.js --steps specialPages
```
with
```
TypeError: Cannot read properties of null (reading 'toLocaleDateString')
    at formatDate (grapher/packages/@ourworldindata/utils/dist/Util.js:960:17)
    at grapher/itsJustJavascript/db/model/Gdoc/GdocBase.js:674:49
    at Array.map (<anonymous>)
    at getMinimalGdocPostsByIds (grapher/itsJustJavascript/db/model/Gdoc/GdocBase.js:668:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async GdocHomepage.loadLinkedDocuments (grapher/itsJustJavascript/db/model/Gdoc/GdocBase.js:507:33)
    at async GdocHomepage.loadState (grapher/itsJustJavascript/db/model/Gdoc/GdocBase.js:626:9)
    at async loadGdocFromGdocBase (grapher/itsJustJavascript/db/model/Gdoc/GdocFactory.js:199:5)
    at async renderFrontPage (grapher/itsJustJavascript/baker/siteRenderers.js:172:30)
    at async SiteBaker.bakeSpecialPages (grapher/itsJustJavascript/baker/SiteBaker.js:353:66)
```

File: baker/batchTagWithGpt.ts [hunk 4]
```diff
@@ -18,14 +18,14 @@ Example: yarn batchTagWithGpt --debug --limit 1
...
-    )
-}
+// export const batchTagWithGpt = async ({
+//     debug,
+//     limit,
+// }: BatchTagWithGptArgs = {}) => {
+//     await db.knexReadonlyTransaction((trx) =>
+//         batchTagChartsWithGpt(trx, { debug, limit })
+//     )
+// }
```
Comment (ikesau) [inline baker/batchTagWithGpt.ts:28]: Commented out?

File: db/tests/basic.test.ts [hunk 5]
```diff
@@ -1,18 +1,15 @@
 #! /usr/bin/env jest
 import sqlFixtures from "sql-fixtures"
 import { dbTestConfig } from "./dbTestConfig.js"
-import { dataSource } from "./dataSource.dbtests.js"
```
Comment (ikesau) [inline db/tests/basic.test.ts:4]: Nice! Hadn't remembered we had these - tested and confirmed working 👍

File: adminSiteServer/adminRouter.tsx [hunk 6]
```diff
@@ -297,19 +309,21 @@ adminRouter.get(`/${EXPLORERS_PREVIEW_ROUTE}/:slug`, async (req, res) => {
...
+    }
+)
 
-adminRouter.get("/datapage-preview/:id", async (req, res) => {
-    const variableId = expectInt(req.params.id)
-    const variableMetadata = await getVariableMetadata(variableId)
-    if (!variableMetadata) throw new JsonError("No such variable", 404)
+getPlainRouteWithROTransaction(
+    adminRouter,
+    "/datapage-preview/:id",
```
Comment (ikesau) [inline adminSiteServer/adminRouter.tsx:320]: I'm getting an error when trying to visit this route: http://localhost:3030/admin/datapage-preview/7216 which requires the server to be restarted

`error: Error: Transaction query already complete, run with DEBUG=knex:tx for more info`

**Review by ikesau — CHANGES_REQUESTED** [review]
Nice! Really amazing effort on all of this - just a little bit more to go!   [🏳️ 🏎️](https://en.wikipedia.org/wiki/Racing_flags#White_flag)

File: adminSiteServer/plainRouterHelpers.tsx (hunk 1 above)
Comment (danyx23) [reply adminSiteServer/plainRouterHelpers.tsx:3]: Yeah I think we could clean that up but it doesn't seem particularly urgent

File: adminSiteServer/testPageRouter.tsx (hunk 2 above)
Comment (danyx23) [reply adminSiteServer/testPageRouter.tsx:441]: Ah nice, thanks for catching that. Fixed on the new top stack PR #3391

File: adminSiteServer/adminRouter.tsx (hunk 6 above)
Comment (danyx23) [reply adminSiteServer/adminRouter.tsx:320]: Thanks for catching that, fixed also in the top PR

**Comment by danyx23** [comment]
### Merge activity

* **Mar 26, 10:04 AM EDT**: @danyx23 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3359).

**Pull request merged**